miden/src/cli/debug.rs

```diff
@@ -0,0 +1,219 @@
+use super::data::{InputFile, ProgramFile};
+use crate::{
+    repl::{print_mem_address, print_stack},
+    VmState,
+};
+use miden::VmStateIterator;
+use rustyline::Editor;
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+#[derive(StructOpt, Debug)]
+#[structopt(name = "Debug", about = "Debug a miden program")]
+pub struct DebugCmd {
+    /// Path to .masm assembly file
+    #[structopt(short = "a", long = "assembly", parse(from_os_str))]
+    assembly_file: PathBuf,
+    /// Path to input file
+    #[structopt(short = "i", long = "input", parse(from_os_str))]
+    input_file: Option<PathBuf>,
+}
+
+impl DebugCmd {
+    pub fn execute(&self) -> Result<(), String> {
+        println!("============================================================");
+        println!("Debug program");
+        println!("============================================================");
+
+        // load program from file and compile
+        let program = ProgramFile::read(&self.assembly_file)?;
+
+        let program_hash: [u8; 32] = program.hash().into();
+        println!("Debugging program with hash {}... ", hex::encode(program_hash));
+
+        // load input data from file
+        let input_data = InputFile::read(&self.input_file, &self.assembly_file)?;
+
+        // fetch the stack and program inputs from the arguments
+        let stack_inputs = input_data.parse_stack_inputs()?;
+        let advice_provider = input_data.parse_advice_provider()?;
+
+        // generate debug trace
+        let mut vm_state_iter = processor::execute_iter(&program, stack_inputs, advice_provider);
+        let mut vm_state = vm_state_iter
+            .next()
+            .ok_or(format!("failed to start iterating vm state"))?
+            .unwrap();
+
+        // initializing readline
+        let mut rl = Editor::<()>::new().expect("Readline couldn't be initialized");
+
+        loop {
+            match rl.readline(">> ") {
+                Ok(command) => match command.as_str() {
+                    // step to the next clock cycle
+                    "!next" => {
```

Comment:
Maybe we should also print the context after calling next? Something like what lldb does:

```
(lldb) next
Process 57693 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = instruction step over
    frame #0: 0x000000010014c0dc miden`main + 8
miden`main:
->  0x10014c0dc <+8>:  stp    x24, x23, [sp, #0xa0]
    0x10014c0e0 <+12>: stp    x22, x21, [sp, #0xb0]
    0x10014c0e4 <+16>: stp    x20, x19, [sp, #0xc0]
    0x10014c0e8 <+20>: stp    x29, x30, [sp, #0xd0]
Target 0: (miden) stopped.
```

I think the output of `!print` would be great for this.

Reply:
I added a `!print` output at the end of all of the clock transition commands.